### PR TITLE
Beef up specialization on non-trivial boolean exprs

### PR DESCRIPTION
**Pull request by abadams**

### CODE REVIEW

**Comment by psuriana** [comment]
Instead of looking for exact match, maybe we should check for "implication" (i.e. whether the condition implies some expr to be true or not) and break into cases as appropriate if it is not too expensive.

**Comment by abadams** [comment]
Done. To make a meaningful test I added simplification rules for testing if a value is within a range. Please check closely. I tried to make a test that was barely true and barely false for each case, without triggering any of the existing rules.

File: test/correctness/specialize.cpp [hunk 1]
```diff
@@ -443,6 +442,58 @@ int main(int argc, char **argv) {
...
+        Param<int> p;
+        Expr test = (p > 73);
+
+        Func f;
+        Var x;
+        f(x) = select(p > 50, im(x, 0), im(0, x));
+        f.specialize(test);
+
+
+        // (p > 73) implies (p > 50), so if the condition h olds (as
```
Comment (psuriana) [inline test/correctness/specialize.cpp]: typo: holds (no space)

File: test/correctness/specialize.cpp [hunk 2]
```diff
@@ -443,6 +442,58 @@ int main(int argc, char **argv) {
...
+
+        Buffer<int> input1(10, 1);
+        Buffer<int> input2(1, 10);
+        im.set(input1);
+        p.set(100);
+        f.realize(10);
+
+        im.set(input2);
+        p.set(-100);
+        f.realize(10);
```
Comment (psuriana) [inline test/correctness/specialize.cpp]: Add a comment on what the test is supposed to do (e.g. throw an error if the condition doesn't simplify, etc.)

File: test/correctness/specialize.cpp [hunk 3]
```diff
@@ -443,6 +442,58 @@ int main(int argc, char **argv) {
...
+
+        // (p <= 73) doesn't tell us anything about (p > 50), so when
+        // the condition doesn't hold, we can make no useful
+        // simplifications. The select remains, so both sides of it
+        // are evaluated, so the image is loaded over the full square.
+        Buffer<int> input2(10, 10);
+        im.set(input2);
+        p.set(-100);
+        f.realize(10);
```
Comment (psuriana) [inline test/correctness/specialize.cpp]: Add a comment on what the test is supposed to do

**Comment by abadams** [comment]
Changes made. PTAL.

**Pull request merged**